### PR TITLE
add SPI NAND Flash driver

### PR DESCRIPTION
**Pull request by rogeryou**
Add SPI NAND block device driver for using SPI NAND Flash like Macronix Flash MX31LF4GE4BC.
This driver is tested on STM32L562E-DK.

### CODE REVIEW

File: drivers/flash/bch.c [hunk 1]
```diff
@@ -0,0 +1,1434 @@
+/*
+ * Generic binary BCH encoding/decoding library
+ *
+ * SPDX-License-Identifier: GPL-2.0
```
Comment (carlescufi) [inline drivers/flash/bch.c:4]: We do not allow GPL-licensed code in Zephyr. You will need to find an alternative implementation or write your own.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: drivers/flash/bch.c (hunk 1 above)
Comment (rogeryou) [reply drivers/flash/bch.c:4]: OK! I will fix this.

Comment (carlescufi) [reply drivers/flash/bch.c:4]: Thanks!

**Comment by GeorgeCGV** [comment]
Can `spi_nand.h` (*new file, added in this PR*) and `spi_nor.h` (*existing header*) not redefine the same definitions?

**Pull request closed without merging**